File: 8-string-to-integer-atoi/8-string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        if(!s.size())
            return 0;
        
        int n=s.length();
        int i{}, sign{1};
        long res{};
        
        for(;i<n && s[i]==' ';++i);
        
        if(i<n && (s[i]=='+' || s[i]=='-'))
            if(s[i++]=='-') sign=-1;
        
        while(i<n && isdigit(s[i])){
            res=res*10 + (s[i]-'0');
            if(sign==1 && res>INT_MAX) return INT_MAX;
            if(sign==-1 && res*sign<INT_MIN) return INT_MIN;
            ++i;
        }
        
        return int(res*sign);
    }
};

File: 8-string-to-integer-atoi/8-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        Boolean sign=false;
        int dg=0;
        int i=0, n=s.length();
        
        while(i<n && s.charAt(i)==' ') ++i;
        if(i<n && (s.charAt(i)=='-' || s.charAt(i)=='+')) sign= (s.charAt(i++)=='-');
        
        while(i<n && s.charAt(i)>='0' && s.charAt(i)<='9'){
            if(dg>Integer.MAX_VALUE/10 || (dg==Integer.MAX_VALUE/10 && (s.charAt(i)-'0')>Integer.MAX_VALUE%10))
                if(sign) return Integer.MIN_VALUE;
                else return Integer.MAX_VALUE;
            
            dg= dg*10 + (s.charAt(i)-'0');
            
            // if(dg>Integer.MAX_VALUE)
            //     if(sign) return Integer.MIN_VALUE;
            //     else return Integer.MAX_VALUE;
            ++i;
        }     
    
        if(sign) dg=-dg;
        return (int)dg;
    }
}